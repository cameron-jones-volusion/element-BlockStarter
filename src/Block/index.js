import styles from './styles';

const defaultConfig = {
    color: {
        background: 'transparent'
    },
    text: 'Default prop'
};

const factory = (
    { React, ElementPropTypes, Components },
    utils,
    { StyleSheet, css },
    globalStyles,
    blockConfig
) => {
    if (!blockConfig.color) {
        blockConfig.color = { background: defaultConfig.color.background };
    }
    const classes = StyleSheet.create(styles(globalStyles, blockConfig));
    const Button = Components.Button.factory(
        { React, ElementPropTypes },
        { StyleSheet, css },
        globalStyles
    );

    const configSpec = {
        color: ElementPropTypes.shape({
            background: ElementPropTypes.color
        }),
        text: ElementPropTypes.string
    };

    const block = class extends React.Component {
        constructor(props) {
            super(props);
            this.state = {
                count: 0
            };
        }

        static defaultProps = defaultConfig;

        click = () => {
            this.setState({ count: ++this.state.count });
        };

        render() {
            return (
                <div className={css(classes.div)}>
                    {this.props.text}{' '}
                    <Button onClick={this.click}>
                        Click me {this.state.count}{' '}
                    </Button>
                </div>
            );
        }
    };

    return {
        block: React.createFactory(block),
        config: configSpec
    };
};

export { factory, defaultConfig };
