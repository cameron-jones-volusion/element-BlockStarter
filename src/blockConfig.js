const configSchema = ElementPropTypes => {
    return {
        text: {
            label: 'Text Content',
            type: ElementPropTypes.string
        }
    };
};

const defaultConfig = {
    text: 'Default prop'
};

export { configSchema, defaultConfig };
