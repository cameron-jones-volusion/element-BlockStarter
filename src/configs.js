export const configSchema = ElementPropTypes => {
    return {
        text: {
            label: 'Text content',
            type: ElementPropTypes.string
        }
    };
};

export const defaultConfig = {
    text: 'Element Starter Block'
};
