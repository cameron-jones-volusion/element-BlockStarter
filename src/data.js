/* eslint-disable-next-line no-unused-vars */
const getDataProps = (utils, props) => {
    return Promise.resolve({});
};

export { getDataProps };
