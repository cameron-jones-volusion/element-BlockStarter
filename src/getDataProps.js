export const getDataProps = () => Promise.resolve();
