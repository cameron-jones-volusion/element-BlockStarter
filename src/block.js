import { defaultConfig } from './blockConfig';

export const factory = React => {
    function StarterBlock(props) {
        return <h1>{props.text}</h1>;
    }
    StarterBlock.defaultProps = defaultConfig;
    return StarterBlock;
};
