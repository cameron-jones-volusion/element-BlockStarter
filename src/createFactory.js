import { factory as blockFactory } from './block';
import { configSchema } from './blockConfig';

const createFactory = ({ React, ElementPropTypes }) => {
    const configSpec = configSchema(ElementPropTypes);

    const block = blockFactory(React);

    return {
        block: React.createFactory(block),
        config: configSpec
    };
};

export default createFactory;
