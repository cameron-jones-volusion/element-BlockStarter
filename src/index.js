import { createFactory } from '@volusion/element-block-utils';
import { StarterBlockFactory } from './block';
import { styles } from './styles';
import { configSchema, defaultConfig } from './configs';
import { getDataProps } from './data';

const factory = createFactory(StarterBlockFactory, styles, configSchema);

export { factory, getDataProps, defaultConfig };
