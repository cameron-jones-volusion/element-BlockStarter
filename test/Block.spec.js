import React from 'react';
import { StyleSheet, css, StyleSheetTestUtils } from 'aphrodite';
import { ElementPropTypes } from '@volusion/element-proptypes';
import * as Components from '@volusion/element-components';
import { configure, shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
configure({ adapter: new Adapter() });

import { factory as createBlock } from '../src/index';

const utils = {};
const globalStyles = {
    typography: {},
    color: {},
    globalComponents: {}
};

describe('The Starter Block', () => {
    beforeEach(() => {
        StyleSheetTestUtils.suppressStyleInjection();
    });

    describe('when there is no custom data', () => {
        it('should render the block with the default content', () => {
            const blockConfig = {};

            const blockSpec = createBlock(
                { React, ElementPropTypes, Components },
                utils,
                { StyleSheet, css },
                globalStyles,
                blockConfig
            );
            const wrapper = shallow(<blockSpec.block />);

            expect(wrapper.dive().props().text).toBe('Default prop');
            expect(wrapper).toMatchSnapshot();
        });
    });

    describe('when given custom data', () => {
        it('should render the block using the custom data', () => {
            StyleSheetTestUtils.suppressStyleInjection();

            const blockConfig = {
                text: 'Hello from props'
            };

            const blockSpec = createBlock(
                { React, ElementPropTypes, Components },
                utils,
                { StyleSheet, css },
                globalStyles,
                blockConfig
            );
            const wrapper = shallow(<blockSpec.block />).setProps(blockConfig);

            expect(wrapper.props().text).toBe('Hello from props');
            expect(wrapper).toMatchSnapshot();
        });
    });

    describe('when mounted with default data', () => {
        it('should match the snapshot', () => {
            const blockConfig = {};

            const blockSpec = createBlock(
                { React, ElementPropTypes, Components },
                utils,
                { StyleSheet, css },
                globalStyles,
                blockConfig
            );
            const wrapper = mount(<blockSpec.block />);

            expect(wrapper).toMatchSnapshot();
        });
    });

    afterEach(() => {
        StyleSheetTestUtils.clearBufferAndResumeStyleInjection();
    });
});
