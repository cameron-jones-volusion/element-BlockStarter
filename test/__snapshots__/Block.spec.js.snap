// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Starter Block when given custom data should render the block using the custom data 1`] = `
<block
  text="Hello from props"
>
  <Component />
</block>
`;

exports[`The Starter Block when mounted with default data should match the snapshot 1`] = `
<bound createElementWithValidation>
  <block>
    <block
      StyleSheet={
        Object {
          "create": [Function],
          "extend": [Function],
          "rehydrate": [Function],
        }
      }
      classes={Object {}}
      css={[Function]}
      text="Default prop"
    >
      <h1>
        Default prop
      </h1>
    </block>
  </block>
</bound createElementWithValidation>
`;

exports[`The Starter Block when there is no custom data should render the block with the default content 1`] = `
<h1>
  Default prop
</h1>
`;
